src/sentence.rs

```diff
@@ -0,0 +1,133 @@
+use std::cell::{RefCell, RefMut};
+use std::rc::Rc;
+
+use conllx::graph::{Node, Sentence};
+use conllx::token::{Token, TokenBuilder};
+use pyo3::class::basic::PyObjectProtocol;
+use pyo3::class::sequence::PySequenceProtocol;
+use pyo3::exceptions;
+use pyo3::prelude::*;
+
+/// Sentence that can be annotated.
+#[pyclass(name=Sentence)]
+pub struct PySentence {
+    inner: Rc<RefCell<Sentence>>,
+}
+
+#[pymethods]
+impl PySentence {
+    /// Construct a new sentence from forms and (optionally) POS tags.
+    ///
+    /// The constructor will throw a `ValueError` if POS tags are
+    /// provided, but the number or tags is not equal to the number of
+    /// tokens.
```

Comment:
The arguments of the `Sentence` type seem to be unknown in python:

~~~python
In [31]: Sentence.__new__
Out[31]: <function Sentence.__new__(*args, **kwargs)>
~~~

Is there a way of making them known?

Reply:
You can add custom signatures, but they don't seem to work for the constructor.

https://pyo3.rs/v0.7.0/function.html#making-the-function-signature-available-to-python

The docstring of the constructor is also not added to help.

Reply:
I briefly checked it doesn't seem to be on their issue tracker yet,

Reply:
Feel free to submit an issue :).

Reply:
I created an issue but my guess is it's a limitation in CPython. `tp_init` and `tp_new` in [typeobject.c](https://github.com/python/cpython/blob/5623ac87bbe5de481957eca5eeae06347612fbeb/Objects/typeobject.c#L3648) are function pointers. The other class methods are in an array of type `PyMethodDef` which holds the field `ml_doc` that gets parsed to create the call signatures. For `__new__` and `__init__` `ml_doc` is discarded and only the functions are assigned to their respective fields in `PyTypeObject`.

Reply:
I tried through the C API (since I have a Python C module lying around anyway), and it indeed does not have any effect adding documentation through a `PyMethodDef` array. Admittedly, the doc for `__new__` also says:

```
Create and return a new object.  See help(type) for accurate signature.
```

Reply:
> `Create and return a new object. See help(type) for accurate signature.`

~~~python
In [1]: from sticker import Sentence                                                                                                                          
In [2]: help(Sentence)          

Help on class Sentence in module builtins:

class Sentence(object)
 |  Sentence that can be annotated.
 |  
 |  Methods defined here:
 |  
 |  __getitem__(self, key, /)
 |      Return self[key].
 |  
 |  __len__(self, /)
 |      Return len(self).
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.
 |  
 |  __repr__(self, /)
 |      Return repr(self).
 |  
 |  __str__(self, /)
 |      Return str(self).
~~~

`help(Sentence)` does not actually provide an accurate signature. I'm not sure how if it's possible at all, but at [typeobject.c:6986](https://github.com/python/cpython/blob/5623ac87bbe5de481957eca5eeae06347612fbeb/Objects/typeobject.c#L6986) the default docs for `__new__` are stored, my guess is at some point they are retrieved and put into the dictionary of the typeobject. Maybe they can be modified there.

Reply:
Let's wait and see what the pyo3 people say.